### PR TITLE
TINY-6505: Add template items to the autocompleter

### PR DESCRIPTION
**Pull request by SimonFc**
Related Ticket: TINY-6505

Description of Changes:
Adding a new template menu item with four new components. This is similar to the dialog panel -> body components setup, allowing for greater customization. As of now it's only being put to use in the autocompleter as a new autocompleter item but we should be able to extend it and use it as an _actual_ menu item in the future.

Not 100% done:
* [x] Fix Css margin issues and make the styles more general

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)
* [x] License headers added on new files (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/oxide/src/less/theme/components/card/card.less [hunk 1]
```diff
@@ -10,25 +10,25 @@
 //
 
 .tox {
-  .tox-card__image {
+  .tox-collection__item-image {
     border-radius: 50%;
```
Comment (lostkeys) [inline modules/oxide/src/less/theme/components/card/card.less:14]: Should be a variable too

File: modules/bridge/src/main/ts/ephox/bridge/api/Menu.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import { ChoiceMenuItem, ChoiceMenuItemInstanceApi, ChoiceMenuItemSpec, createChoiceMenuItem } from '../components/menu/ChoiceMenuItem';
+import { CardMenuItemSpec, CardMenuItem, createCardMenuItem, CardMenuItemInstanceApi, CardItem } from '../components/menu/CardMenuItem';
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/api/Menu.ts:2]: I noticed this when reviewing the mentions code, but this is the wrong location to export this IIUC. Any autocompleter items should be exposed via the `InlineContent` module, not menus.

Anything thing that's missing, is the specs haven't been exposed via the appropriate `Public` module in bridge (eg PublicMenu, PublicInlineContent, etc...).

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Autocomplete.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { FieldSchema, ValueSchema } from '@ephox/boulder';
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/components/menu/Autocomplete.ts:1]: This is almost definitely the wrong place for this module, or it needs to be renamed to something other than "autocomplete". All autocompleter modules should be in `components/content`.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import { FieldSchema, ValueSchema } from '@ephox/boulder';
+import { Fun, Result } from '@ephox/katamari';
+import { Autocomplete, AutocompleteSpec } from './Autocomplete';
+import { CommonMenuItem, commonMenuItemFields, CommonMenuItemInstanceApi, CommonMenuItemSpec } from './CommonMenuItem';
+import { Container, ContainerSpec, itemSchema } from './Container';
+import { Description, DescriptionSpec } from './Description';
+import { Image, ImageSpec } from './Image';
+
+export type CardItemsSpec =
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts:9]: This is missing `Menu` in the name, so it's inconsistent. It should probably be `CardMenuItemsSpec`.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/structure/ItemSlices.ts [hunk 5]
```diff
@@ -88,6 +89,38 @@ const renderDownwardsCaret = (icons: IconProvider): AlloySpec => ({
...
+  },
+  components
+});
+
+const renderImage = (src: string, width: Optional<string>, height: Optional<string>, alt: Optional<string>): AlloySpec => ({
+  dom: {
+    tag: 'img',
+    classes: [ ItemClasses.imageClass ],
+    styles: {
+      'vertical-align': 'top'
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/item/structure/ItemSlices.ts:105]: Why is this here? This doesn't seem like a common thing we'd want and this should probably instead be in the css.

File: modules/bridge/src/main/ts/ephox/bridge/api/Menu.ts (hunk 2 above)
Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/api/Menu.ts:2]: After having read more of the PR, it seems card is meant to be a new generic component for menu items. As such there's some other changes that need to be made:
- Needs adding to the relevant menu specs/logic (eg `NestedMenuItemContents`)
- `autocomplete` is a confusing property on something that's meant to be generic, so it needs a new name. I don't have any suggestions atm sorry.
- Needs to be added to the registry, as something that can be registered.
  - To match this, the autocompleter types need to handle a reference (name) and the logic in silver probably also needs to be updated to lookup names from the registry.
- `CardItemsSpec` probably should be exported, so it can be used when generating the card spec.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import { FieldSchema, ValueSchema } from '@ephox/boulder';
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts:1]: This really isn't a menu, it should probably be moved to a subdirectory, eg: `components/menu/card/Container`. The same applies to the other card item sub components.

Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts:1]: Also maybe this should be called `CardContainer` to avoid confusion?

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import { Result } from '@ephox/katamari';
+import { CardItem } from '../../api/Menu';
+import { autocompleteSchema } from './Autocomplete';
+import { CardItemsSpec } from './CardMenuItem';
+import { descriptionSchema } from './Description';
+import { imageSchema } from './Image';
+
+export interface ContainerSpec {
+  type: 'container';
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts:10]: This name is going to clash if we ever want a more generic container (remember types aren't namespaced, they are global). I'd suggest renaming this to `cardcontainer` or something more specific.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/structure/ItemSlices.ts [hunk 8]
```diff
@@ -88,6 +89,38 @@ const renderDownwardsCaret = (icons: IconProvider): AlloySpec => ({
...
+    },
+    attributes: {
+      src,
+      ...width.map((width) => ({ width })).getOr({}),
+      ...height.map((height) => ({ height })).getOr({}),
+      ...alt.map((alt) => ({ alt })).getOr({})
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/item/structure/ItemSlices.ts:111]: I mentioned this on the mentions PR, but without an `alt` this will cause an accessibility issue. So should this just default to using `alt=''` so it's treated as a decorative image?

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    case 'description':
+      return renderDescription(item.text);
+
+    case 'autocomplete':
+      const matchText = Obj.get(extras, 'autocompleteMatchText').getOr('');
+      return renderHtml(replaceText(item.text, matchText));
+  }
+});
+
+export function renderCardMenuItem(
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:40]:
```suggestion
export const renderCardMenuItem = (
```

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  direction?: 'vertical' | 'horizontal';
+  items: CardItemsSpec[];
+}
+
+export interface Container {
+  type: 'container';
+  direction: 'vertical' | 'horizontal';
+  items: CardItem[];
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts:18]: As discussed in the call, we should possibly rename this to `ContainerItem` or something more generic.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Autocomplete.ts (hunk 3 above)
Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Autocomplete.ts:1]: For the name, we discussed that maybe this should be called `title`?

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts (hunk 4 above)
Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts:9]: Update: Ignore this, as we said we should rename this to be more generic (see the comment on `Container.ts`).

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+import { Arr, Obj, Optional } from '@ephox/katamari';
+import { UiFactoryBackstageShared } from 'tinymce/themes/silver/backstage/Backstage';
+import { renderItemDomStructure } from 'tinymce/themes/silver/ui/menus/item/structure/ItemStructure';
+import ItemResponse from '../ItemResponse';
+import { renderContainer, renderDescription, renderHtml, renderImage } from '../structure/ItemSlices';
+import { replaceText } from './AutocompleteMenuItem';
+import { buildData, renderCommonItem } from './CommonMenuItem';
+
+interface CardExtras {
+  autocompleteMatchText?: string;
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:19]: Maybe just call this match text, so it's not tied to the autocompleter specifically?

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+});
+
+export function renderCardMenuItem(
+  spec: Menu.CardMenuItem,
+  itemResponse: ItemResponse,
+  sharedBackstage: UiFactoryBackstageShared,
+  extras: CardExtras
+) {
+  const getApi = (component: AlloyComponent): Menu.CardMenuItemInstanceApi => ({
+    isDisabled: () => Disabling.isDisabled(component),
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:47]: This will only disable the "card" component, not the child components (eg title, description) I believe. Will that cause issues?

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/menu/SingleMenu.ts [hunk 13]
```diff
@@ -104,25 +105,46 @@ export const createAutocompleteItems = (
...
+          );
+
+        case 'cardmenuitem':
+          return BridgeMenu.createCardMenuItem(item).fold(
+            MenuUtils.handleError,
+            (d) => Optional.some(MenuItems.card(
+              {
+                ...d,
+                // Intercept action
+                onAction: (_api) => onItemValueHandler(d.value, d.meta)
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/menu/SingleMenu.ts:122]: As discussed, we should probably still call the original `onAction` still if it was defined and then fire the autocomplete on action.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts (hunk 6 above)
Comment (SimonFc) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts:1]: All the Panel sub components are placed directly in `components/dialog/` so to keep it consistent with that, should everything be in `components/menu/` after all?

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/structure/ItemSlices.ts (hunk 5 above)
Comment (SimonFc) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/item/structure/ItemSlices.ts:105]: Correct, commited this by mistake when experimenting.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts (hunk 12 above)
Comment (SimonFc) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:47]: There's no child component that got any interactivity so do they need to be disabled? Not sure we would have any either since it would be inside an autocompleter/menu item?

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts (hunk 7 above)
Comment (SimonFc) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts]: I renamed this to `cardcontainer`

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/autocomplete/AutocompleterTest.ts [hunk 14]
```diff
@@ -563,6 +606,48 @@ UnitTest.asynctest('Editor Autocompleter test', (success, failure) => {
...
+                    store.add('euro:AutocompleterContents->onAction');
+                  },
+                  items: [
+                    { type: 'image', src: getGreenImageDataUrl(), classes: [ 'my_autocompleter_avatar_class' ] },
+                    {
+                      type: 'cardcontainer',
+                      direction: 'vertical',
+                      items: [
+                        { type: 'title', text: item.text, classes: [ 'title_class' ], name: 'my_title_to_highlight' },
+                        { type: 'description', text: `Description for ${item.text}`, classes: [ 'description_class' ] }
```
Comment (SimonFc) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/autocomplete/AutocompleterTest.ts:636]: After our latest design decision to allow styling with classes, I don't think it makes sense to have more than one text-component. If you need something like a description with smaller text you can achieve that via a class so I'll probably just remove this.

Comment (lostkeys) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/autocomplete/AutocompleterTest.ts:636]: I agree!

Comment (lostkeys) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/autocomplete/AutocompleterTest.ts:636]: How do you define which field that shows the autocompleter matches...?

Comment (SimonFc) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/autocomplete/AutocompleterTest.ts:636]: The way it works with these cards is that the autocompleter now got a `highlight` option (see a few lines above) which will highlight the item(s) with a matching `name`-property. In this test it's the component with the "my_title_to_highlight" name. We had a short discussion about alternatives but this is probably the simplest solution.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts (hunk 6 above)
Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts]: That only ever has 1 top-level component though, this has 4 now so we need to separate it somewhat, otherwise it'll get confusing about what's a sub vs top-level component. Ideally, the dialog stuff should be broken up as well IMO, but that's out of scope.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts (hunk 12 above)
Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:52]: While that's true atm, I'm concerned it'd easily get missed if we do add something.

File: modules/bridge/src/main/ts/ephox/bridge/api/PublicMenu.ts [hunk 15]
```diff
@@ -29,5 +30,11 @@ export {
   ContextMenuApi,
   ContextMenuContents,
   ContextMenuItem,
-  ContextSubMenu
+  ContextSubMenu,
+
+  CardMenuItemSpec,
+  CardMenuItem,
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/api/PublicMenu.ts:36]: These shouldn't be here, as it's an internal-only spec (likewise with ContainerItem below). This module should only have public types that are used when registering the component by the APIs.

Also, should this have `CardContainerSpec` and all those other relevant types exposed?

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts [hunk 16]
```diff
@@ -0,0 +1,25 @@
+import { FieldSchema, ValueSchema } from '@ephox/boulder';
+import { Result } from '@ephox/katamari';
+
+export interface DescriptionSpec {
+  type: 'description';
+  text: string;
+  classes?: string[];
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts:7]: Should we have classes here? I thought we were specifically avoiding allowing integrators to use classes, as it caused support issues.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Image.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
+import { FieldSchema, ValueSchema } from '@ephox/boulder';
+import { Optional, Result } from '@ephox/katamari';
+
+export interface ImageSpec {
+  type: 'image';
+  src: string;
+  classes?: string[];
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/components/menu/Image.ts:7]: Likewise, classes is normally an attribute we don't want to expose here.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Title.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
+import { FieldSchema, ValueSchema } from '@ephox/boulder';
+import { Optional, Result } from '@ephox/katamari';
+
+export interface TitleSpec {
+  type: 'title';
+  text: string;
+  name?: string;
+  classes?: string[];
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/components/menu/Title.ts:8]: Another place where we shouldn't be including classes.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts (hunk 16 above)
Comment (SimonFc) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts:7]: We had a "Quality of Life Improvements" discussion about this and we considered two alternatives when it comes to styling: 1) add style attributes for a set of safe styles or 2) allow classes and document what kind of safe styles (non-layout) that an integrator may use. We thought alternative 2 would be the better choice here. Anything you would like to add @lostkeys or @metricjs?

Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts:7]: If we're going to do that it needs to be a separate PR and not something partially done here.

Comment (lostkeys) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts:7]: @lnewson We decided classes on these elements is a reasonable compromize between a very verbose list of attributes props and allowing the integrator to fully custom html and css. Since the class is applied to know elements and in a structured way, we can much better anticipate what integrators might do than if they could throw any complex HTML Dom tree inside our UI. Its mainly changes our reset and layout css that could break customers integrations, but when targeted so specifically as we do here, it's not as "dangerous".

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Image.ts (hunk 17 above)
Comment (lostkeys) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Image.ts:7]: See https://github.com/tinymce/tinymce/pull/6149#discussion_r515868104

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Title.ts (hunk 18 above)
Comment (lostkeys) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Title.ts:8]: See https://github.com/tinymce/tinymce/pull/6149#discussion_r515868104

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts (hunk 16 above)
Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts:7]: That might be true now, but what about in future if we add more components? I'm probably okay with adding it on the `container`, but it shouldn't be on `image`, `title` or `description` as those could end up being used elsewhere potentially. It'd also be. very confusing to users to have somethings that allow classes, but other things that don't.

File: modules/bridge/src/main/ts/ephox/bridge/api/Menu.ts [hunk 19]
```diff
@@ -42,5 +43,12 @@ export {
...
   ContextMenuItem,
-  ContextSubMenu
+  ContextSubMenu,
+
+  createCardMenuItem,
+  CardMenuItem,
+  CardMenuItemSpec,
+  CardMenuItemInstanceApi,
+  ContainerItemSpec,
+  ContainerItem
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/api/Menu.ts:53]: Should this have all the other container types exposed? As they can be used, so I believe the types should be exposed.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts (hunk 16 above)
Comment (lostkeys) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts:7]: Sorry but it's the container we _don't_ want users to add classes to. It was unfortunate you weren't in the meeting. The reason for this is to be able to style all the things needed for the flexibility of this menu item. Such as setting width, height and alignment of images, setting font sizes, colors, text transforms etc on text nodes. These things cannot go in the skin as we must support different styles for different autocompleter menus, as well as cases where a menu item can contains multiple images. The other solution was to use a styles property but we deemed classes be the more robust option. 

Here is an image showing an example what we want to be able to do
Untitled-2

Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts:7]: Fair enough, in that case though then the new components shouldn't be generic and need to be more specific, as we don't want `classes` on something generic.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+  CardContainer |
+  Image |
+  Description |
+  Title;
+
+export interface CardMenuItemInstanceApi extends CommonMenuItemInstanceApi { }
+
+export interface CardMenuItemSpec extends Omit<CommonMenuItemSpec, 'text'> {
+  type: 'cardmenuitem';
+  ariaLabel?: string;
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts:25]: Why do we have `ariaLabel` here, as we don't include that on any other menu item? If we do need this, I believe this should instead be either be `label` or `tooltip` to keep consistent with the rest of the bridge spec/api.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+import { UiFactoryBackstageShared } from 'tinymce/themes/silver/backstage/Backstage';
+import { renderItemDomStructure } from 'tinymce/themes/silver/ui/menus/item/structure/ItemStructure';
+import ItemResponse from '../ItemResponse';
+import { renderContainer, renderDescription, renderHtml, renderImage } from '../structure/ItemSlices';
+import { replaceText } from './AutocompleteMenuItem';
+import { buildData, renderCommonItem } from './CommonMenuItem';
+
+export interface CardExtras {
+  itemBehaviours?: Array<Behaviour.NamedConfiguredBehaviour<Behaviour.BehaviourConfigSpec, Behaviour.BehaviourConfigDetail>>;
+  title: { // Extras specific to title components
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:20]: Maybe move this comment to the line above? as this really applies to the whole title block.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts (hunk 16 above)
Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts:7]: To clarify, when I say more specific, I mean calling them `cardtitle` instead of `title` for example.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+
+    case 'image':
+      return renderImage(item.src, item.classes, item.alt);
+
+    case 'description':
+      return renderDescription(item.text, item.classes);
+
+    case 'title':
+      // Only highlight targeted titles
+      const shouldHighlight = item.name.exists((name) => Arr.contains(extras.title.highlight, name));
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:39]: This appears to be missing from the `description`?  When we discussed on Thursday last week it was my understanding that should also be able to be highlighted if needed.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/menu/SingleMenu.ts [hunk 23]
```diff
@@ -97,32 +98,60 @@ export const createAutocompleteItems = (
...
+
+        case 'cardmenuitem':
+          return BridgeMenu.createCardMenuItem(item).fold(
+            MenuUtils.handleError,
+            (d) => Optional.some(MenuItems.card(
+              {
+                ...d,
+                // Intercept action
+                onAction: (api) => {
+                  onItemValueHandler(d.value, d.meta);
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/menu/SingleMenu.ts:124]: When we discussed this last week, we said this should be the opposite order. As it'd be odd to have the generic handler fire before the handler that's specific to the. menu item.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts (hunk 20 above)
Comment (SimonFc) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts:25]: I thought "text" made little sense since it wont ever be put in the dom as a `tox-collection__item-label` item, which other menu items do, but I can change that.

Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts:25]: Yeah, I'd agree `text` doesn't make sense since it won't ever be visible. However `ariaLabel` isn't used anywhere else, that was more the point I was trying to make sorry.

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts (hunk 22 above)
Comment (SimonFc) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:39]: I'll remove the description component, this [comment](https://github.com/tinymce/tinymce/pull/6149#discussion_r515174905)

Comment (lnewson) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:39]: Sounds good, I assume you'll just go with a `cardtext` or something component instead?

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts (hunk 20 above)
Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts:25]: Speaking of this, does it make sense to include `shortcut` as well, given that also won't ever be visible?

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts (hunk 16 above)
Comment (lostkeys) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts:7]: Ok I (think I) see, I have no problem with renaming the component to be `cardtext` or something similar to avoid people thinking they can add classes to any `title` property they find in TinyMCE. That makes total sense.

Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts:7]: Yeah, that's all I meant if we want to allow classes to be applied for this use case 👍

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts (hunk 22 above)
Comment (SimonFc) [reply modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:39]: Yeah I thought about what we could call the `title` component instead and `cardtext` sounds good and generic

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts (hunk 20 above)
Comment (SimonFc) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts]: Great that you saw that, `shortcut` should be excluded as well.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts (hunk 6 above)
Comment (SimonFc) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts]: I moved the subcomponents to a `menu/card` subdirectory as I think it makes sense have these in the same place.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts (hunk 16 above)
Comment (SimonFc) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Description.ts]: I have renamed the components to `cardtext` and `cardimage` (Description has been removed)

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts (hunk 10 above)
Comment (SimonFc) [reply modules/bridge/src/main/ts/ephox/bridge/components/menu/Container.ts]: I renamed the type to `ContainerItem` earlier but (after some thinking) I can't see how it should be called container. In the case of [Panel](https://github.com/tinymce/tinymce/blob/9f1f378779b2a68d8a4b577f341461616d408ac3/modules/bridge/src/main/ts/ephox/bridge/components/dialog/Panel.ts#L35), the items-type is called `BodyComponent` (which may be a Label or Checkbox etc.).

I changed this again and restructured to reflect the structure of `BodyComponent`. See this [commit](https://github.com/tinymce/tinymce/pull/6149/commits/cd665466a76e7d6fc5b916e7919d656815504aba). Do you think it makes more sense now after the other changes we have made, and how it has become more `Card`-specific overall? if not, I can revert this commit.

File: modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+
+export interface CardMenuItemSpec extends Omit<CommonMenuItemSpec, 'text' | 'shortcut'> {
+  type: 'cardmenuitem';
+  label?: string;
+  items: CardItemSpec[];
+  onSetup?: (api: CardMenuItemInstanceApi) => (api: CardMenuItemInstanceApi) => void;
+  onAction?: (api: CardMenuItemInstanceApi) => void;
+}
+
+export interface CardMenuItem extends Omit<CommonMenuItem, 'text'> {
```
Comment (lnewson) [inline modules/bridge/src/main/ts/ephox/bridge/components/menu/CardMenuItem.ts:17]: Should this also have `shortcut` omitted?

**Review by lnewson — APPROVED** [review]
LGTM, thanks for all the additional changes and the new feature! :tada:

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/structure/ItemSlices.ts [hunk 25]
```diff
@@ -88,6 +90,34 @@ const renderDownwardsCaret = (icons: IconProvider): AlloySpec => ({
...
+  dom: {
+    tag: 'div',
+    classes: [
+      ItemClasses.containerClass
+    ].concat([
+      container.direction === 'vertical' ? ItemClasses.containerColumnClass : ItemClasses.containerRowClass
+    ]).concat([
+      container.align === 'left' ? ItemClasses.containerAlignLeftClass : ItemClasses.containerAlignRightClass
+    ]).concat([
+      container.valign === 'top' ? ItemClasses.containerValignTopClass :
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/item/structure/ItemSlices.ts:103]: This is nested ternaries, which are normally banned. Can this be rewritten to remove the nested ternary?

- [ ] Rewrite to avoid nested ternaries if possible.

**Review by JamesToohey — APPROVED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+  sharedBackstage: UiFactoryBackstageShared,
+  extras: CardExtras
+) => {
+  const getApi = (component: AlloyComponent): Menu.CardMenuItemInstanceApi => ({
+    isDisabled: () => Disabling.isDisabled(component),
+    setDisabled: (state: boolean) => {
+      Disabling.set(component, state);
+
+      // Disable sub components
+      Selectors.all('*', component.element).forEach((elm) => {
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/main/ts/ui/menus/item/build/CardMenuItem.ts:57]: I noticed this when talking about something else, but this is using the native `forEach`. This should use `Arr.each` instead. It should also probably be using `SelectorFilter.descendants(component.element, '*')` for better clarity.

**Review by metricjs — APPROVED** [review]
I also pulled the branch and played with a simple demo based on one of the test cases, and it seemed to work well!

File: modules/bridge/src/main/ts/ephox/bridge/components/content/Autocompleter.ts [hunk 27]
```diff
@@ -35,6 +36,7 @@ export interface AutocompleterSpec {
   fetch: (pattern: string, maxResults: number, fetchOptions: Record<string, any>) => Promise<AutocompleterContents[]>;
   onAction: (autocompleterApi: AutocompleterInstanceApi, rng, value: string, meta: Record<string, any>) => void;
   maxResults?: number;
+  highlight?: string[];
```
[the remark this answers is not in the document]
Comment (metricjs) [reply modules/bridge/src/main/ts/ephox/bridge/components/content/Autocompleter.ts:39]: Talked about this with Lee a little and we wondered if `highlightOn` would make a little more sense? I don't really mind either way though, and we can always tweak it over the next week if we do think of something better.

[the remark this answers is not in the document]
Comment (lnewson) [reply modules/bridge/src/main/ts/ephox/bridge/components/content/Autocompleter.ts:39]: The other option we mentioned was `highlightItems` if that's any better as well?

[the remark this answers is not in the document]
Comment (SimonFc) [reply modules/bridge/src/main/ts/ephox/bridge/components/content/Autocompleter.ts]: I went for `highlightOn` (`highlightItems` would have worked as well)

**Review by metricjs — APPROVED** [review]
